### PR TITLE
feat(fbw): improved speed dependent roll law model

### PR DESCRIPTION
**Pull request by IbrahimK42**
## Summary of Changes

The roll law with speed independent aileron elasticity table showed instability for higher frame rates. The implemented time compression solved the issue.  The roll law in this PR depends on calibrated airspeed and uses the speed independent aileron elasticity table.

## Testing instructions

1. Roll maneuvres with IAS (150kts, 180kts, 220kts, 250kts) at 5000ft AGL  with SimRate = 1 and higher SimRates
1. Roll maneuvres with IAS (190kts, 180kts, 220kts, 260kts, 300kts) at 10000ft AGL with SimRate = 1 and higher SimRates
1. Roll maneuvres with IAS (230kts, 260kts, 300kts) at 30000ft AGL with SimRate = 1 and higher SimRates

Questions:
- Do you notice any instability in higher sim rates?
- How is the roll performance?

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by aguther — APPROVED** [review]
I've tested the new roll law model with normal flight and also with time compression up to 4x and it seems stable. I will test the coming days (as time allows) with other conditions and flights etc.

From my side it's ready to go to QA.

**Comment by ShadowFrost3** [comment]
Quality Assurance Trainee Report


Object of testing: #5842 
Tier of Testing: .5
Date: 19/09/2021

Testing Process: Flew a complete normal flight and tested various turns at varying altitudes and speed. No abnormalities to report, bank may be slightly aggressive at low altitudes. (personal perception, awaiting further information) 
Negatives: 
Testing Results: N/A - See below

Conclusions: More information requested as to the expected roll parameters and FBW logic. Nothing game breaking seen, but I would like to verify bank limits and roll degrees /s allowed for given airspeed/altitudes/conditions. Feel free to message on discord, I will edit this comment in time.

**Comment by IbrahimK42** [comment]
> Quality Assurance Trainee Report
> 
> Object of testing: #5842
> Tier of Testing: .5
> Date: 19/09/2021
> 
> Testing Process: Flew a complete normal flight and tested various turns at varying altitudes and speed. No abnormalities to report, bank may be slightly aggressive at low altitudes. (personal perception, awaiting further information)
> Negatives:
> Testing Results: N/A - See below
> 
> Conclusions: More information requested as to the expected roll parameters and FBW logic. Nothing game breaking seen, but I would like to verify bank limits and roll degrees /s allowed for given airspeed/altitudes/conditions. Feel free to message on discord, I will edit this comment in time.

The roll law model generates from pilot roll command (max. +-15 deg/s) a bank angle command and limits here the bank angle to +-67 deg (for all flight conditions). The Aileron Elasticity Table in flight_model.cfg caused the aileron deflection to depend on the speed. This aileron_elasticity_table is in this PR set to speed independent (using 0:1, 400:1). The gains of the bank angle controller were now designed in such a way that, that depending on the flight speed, the transmission behaviour between roll rate command and measured roll rate is a transmission behaviour 1st order delay without overshoot (PT1-behaviour). This then also prevents a rollback of the bank angle at higher speeds. If you have any further questions, I would be happy to answer them :)

**Review by lukecologne — APPROVED** [review]

**Review by lolisec — APPROVED** [review]

**Comment by ShadowFrost3** [comment]
Quality Assurance Trainee Report


Object of testing: #5842 
Tier of Testing: 2
Date: 19/09/2021

Testing Process: Flew two complete normal flights and tested various turns, direct intercepts, and heading intercepts at varying altitudes and speeds relevant to normal operation. (140, 180, 250, 280, Mach cruises)  Tested throughout climb and 5,000, 10,000 and cruise (30,000 & 37,000).
Auto pilot intercepts heading and turns without issue and gradually reduces turn limit from 15 degrees to lower at higher speed/altitude. 
Pilot roll command appears normal with a reduction in roll rate approaching either maximum of 45 or 67 degrees given conditions to avoid overbanking.  
Verified overspeed Roll law. Maximum pilot bank was ~40 degrees and reduced to 0 with neutral sidestick input. 
Verified AOA Roll law. Reduced bank to 45 degrees and reduced to 0 with neutral sidestick input. 

Aircraft is good to go on sim rate 1. 

Above sim rate 1 and to the max simrate (unsure how this is currently defined as was recently changed, I believe 4x is max)
Aircraft handles autopilot turns well - no noticeable issues.
Non-autopilot max bank is handled well at altitudes of 5,000 and 10,000. 
Max bank (autopilot off) into overspeed does not break the aircraft from 0-10,000ft. 

Non autopilot, max bank issues encounter at ~30,000FT and above. - Failed
Max bank into overspeed (mach .78 and 35,000ft) enters the aircraft into oscillation and unrecoverable state
Three attempts on dev build to recreate the same issue failed. 
Stable build was not tested. 
Media below


 
Negatives: 
Non autopilot, bank issues encounter at ~30,000FT and above. 
Max bank turn at 35,000ft mach .78, aircraft begins a leftward spiral towards overspeed limit. When overspeed limit and corresponding law is applied the aircraft continuously oscillates and enters a unrecoverable attitude. 

Testing Results: Failed - Many aspects are good but this isn't a small issue (albeit however unlikely to be encountered). 

Conclusions: I may have tested a fair amount more than needed... 

The problem
https://streamable.com/5fsc23
Dev build test (no problem)
https://streamable.com/shzsjj

**Comment by IbrahimK42** [comment]
> Quality Assurance Trainee Report
> 
> Object of testing: #5842
> Tier of Testing: 2
> Date: 19/09/2021
> 
> Testing Process: Flew two complete normal flights and tested various turns, direct intercepts, and heading intercepts at varying altitudes and speeds relevant to normal operation. (140, 180, 250, 280, Mach cruises) Tested throughout climb and 5,000, 10,000 and cruise (30,000 & 37,000).
> Auto pilot intercepts heading and turns without issue and gradually reduces turn limit from 15 degrees to lower at higher speed/altitude.
> Pilot roll command appears normal with a reduction in roll rate approaching either maximum of 45 or 67 degrees given conditions to avoid overbanking.
> Verified overspeed Roll law. Maximum pilot bank was ~40 degrees and reduced to 0 with neutral sidestick input.
> Verified AOA Roll law. Reduced bank to 45 degrees and reduced to 0 with neutral sidestick input.
> 
> Aircraft is good to go on sim rate 1.
> 
> Above sim rate 1 and to the max simrate (unsure how this is currently defined as it was recently changed, I believe 4x is max)
> Aircraft handles autopilot turns well - no noticeable issues.
> Non-autopilot max bank is handled well at altitudes of 5,000 and 10,000.
> Max bank (autopilot off) into overspeed does not break the aircraft from 0-10,000ft.
> 
> Non autopilot, max bank issues encounter at ~30,000FT and above. - Failed
> Max bank into overspeed (mach .78 and 35,000ft) enters the aircraft into oscillation and unrecoverable state
> Three attempts on dev build to recreate the same issue failed.
> Stable build was not tested.
> Media below
> 
> Bug recreation.
> Time Accel x4 (or max)
> Altitude of 35,000ft, mach .78, heading mode.
> Disconnect autopilot. Auto throttles remain on.
> Initiate full left input on the side stick and hold.
> When aircraft encounter overspeed law aircraft will become unstable
> No attempts to recover were successful
> 
> Negatives:
> Non autopilot, bank issues encounter at ~30,000FT and above.
> Max bank turn at 35,000ft mach .78, aircraft begins a leftward spiral towards overspeed limit. When overspeed limit and corresponding law is applied the aircraft continuously oscillates and enters a unrecoverable attitude.
> 
> Testing Results: Failed - Many aspects are good but this isn't a small issue (albeit however unlikely to be encountered).
> 
> Conclusions: I may have tested a fair amount more than needed...
> 
> The problem
> https://streamable.com/5fsc23
> Dev build test (no problem)
> https://streamable.com/shzsjj

Thank you for your detailed testing. The PR with a more stable time compression 
#5863  solves the issue.

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 <!--  1. [Changed Area] Title of changes - @github username (Name)  -->
 
 ## 0.8.0
+1. [FBW] Improved speed dependent roll law model - @IbrahimK42 (IbrahimK42)
```
Comment (davidwalschots) [inline .github/CHANGELOG.md:8]: **nitpick**: Changelog entries should go to the bottom of the list 😀.

**Comment by aguther** [comment]
> 
> Thank you for your detailed testing. The PR with a more stable time compression
> #5863 solves the issue.

I’m not sure about this. The time compression improvement is already on development. The change that Lukas has done is only to change the algorithm how sim rate gets reduced - and it’s doing that less aggressive than it was…

**Comment by IbrahimK42** [comment]
> > Thank you for your detailed testing. The PR with a more stable time compression
> > #5863 solves the issue.
> 
> I’m not sure about this. The time compression improvement is already on development. The change that Lukas has done is only to change the algorithm how sim rate gets reduced - and it’s doing that less aggressive than it was…

I tested it for high speed and 4x with overbanking and it is really oscillating.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (IbrahimK42) [reply .github/CHANGELOG.md:8]: Ups, it was my first PR ever :) Will pay attention next time

**Comment by aguther** [comment]
The PR https://github.com/flybywiresim/a32nx/pull/5863 will per default also reduce simulation rate when being in abnormal situation (large bank / pitch and any protection active). This should avoid the mentioned scenario.

**Comment by aguther** [comment]
For me personally it's good to go. This new roll law is on experimental and we didn't see any new reports. If the issue appears again we need to further think what we can do. Right now it seems it's not an issue anymore.

**Pull request merged**